src/com.rs

```diff
@@ -11,44 +11,90 @@ use winapi::{ctypes::c_void, um::unknwnbase::IUnknown, Interface};
 pub struct ComPtr<T: Interface>(*mut T);
 
 impl<T: Interface> ComPtr<T> {
+    /// Creates a null ComPtr.
     pub fn null() -> Self {
         ComPtr(ptr::null_mut())
     }
 
+    /// Create a ComPtr from a raw pointer. This will call AddRef on the pointer.
+    ///
+    /// # Safety
+    ///
+    /// - if `raw` is not null, it must be a valid pointer to a COM object that implements T.
     pub unsafe fn from_raw(raw: *mut T) -> Self {
         if !raw.is_null() {
-            (&*(raw as *mut IUnknown)).AddRef();
+            (*(raw as *mut IUnknown)).AddRef();
         }
         ComPtr(raw)
     }
 
+    /// Returns true if the inner pointer is null.
     pub fn is_null(&self) -> bool {
         self.0.is_null()
     }
 
+    /// Returns the raw inner pointer. May be null.
     pub fn as_ptr(&self) -> *const T {
         self.0
     }
 
+    /// Returns the raw inner pointer as mutable. May be null.
     pub fn as_mut_ptr(&self) -> *mut T {
         self.0
     }
 
-    pub fn mut_void(&mut self) -> *mut *mut c_void {
-        &mut self.0 as *mut *mut _ as *mut *mut _
+    /// Returns a mutable reference to the inner pointer casted as a pointer to c_void.
+    ///
+    /// This is useful when D3D functions initialize objects by filling in a pointer to pointer
+    /// by taking `void**` as an argument.
+    ///
+    /// # Safety
+    ///
+    /// - Any modifications done to this pointer must result in the pointer either:
+    ///   - being set to null
+    ///   - being set to a valid pointer to a COM object that implements T
+    pub unsafe fn mut_void(&mut self) -> &mut *mut c_void {
+        // SAFETY: We must first get a reference pointing to our internal pointer
+        // and only then cast it. As if we cast it, then take a reference, we would
+        // end up with a reference to a temporary.
+        let refer: &mut *mut T = &mut self.0;
+        let void: *mut *mut c_void = refer.cast();
```

Comment:
In retrospect, this was incorrect, and sets void to `self.0 as *mut *mut c_void` instead of `&mut self.0 as *mut *mut T as *mut *mut c_void`, because .cast applies to the primitive pointer `*mut T` by reference.

Reply:
In addition to the problem that @kdashg points out, there's another issue here: `&mut *void` is actually a reference to a `*mut c_void` temporary in this method, whose lifetime is incorrectly inferred through the pointer deref. (because `rustc` trusts us to get the lifetime right). This becomes insta-UB. 😓

I don't think it makes sense to use mutable references here, TBH. I'll fix this up in https://github.com/gfx-rs/wgpu/pull/3987, which vendors in this crate (and will likely result in this repository being archived, for the time being).

CC @jrmuizel; I think he'd want to know about these issues.